v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+
+            if do_set_disconnect:
+                # Change state and notify tasks waiting on disconnect
+                async def set_disconnected():
+                    logger.debug("Client State: DISCONNECTED")
+                    this._connected = False
+                    async with this.disconnected_cond:
+                        this.disconnected_cond.notify_all()
+
+                f = asyncio.run_coroutine_threadsafe(set_disconnected(), this._loop)
+                # Need to wait for this one to finish since we don't want to let another
+                # Paho handler invoke until we know the connection state has been set.
+                f.result()
+
+            if do_stop_network_loop:
+
+                async def stop_network_loop():
+                    logger.debug("Stopping Paho network loop due to unexpected disconnect")
+                    client.loop_stop()
+
+                # NOTE: This coroutine can't finish execution until after this handler exits
+                # since this handler itself is running on the network loop.
+                # This means this MUST be the last task that gets scheduled in this handler,
+                # since .loop_stop() will block until this handler finishes running.
+                asyncio.run_coroutine_threadsafe(stop_network_loop(), this._loop)
+
+        def on_subscribe(client, userdata, mid):
+            pass
+
+        def on_unsubscribe(client, userdata, mid):
+            pass
+
+        mqtt_client.on_connect = on_connect
+        mqtt_client.on_disconnect = on_disconnect
+        mqtt_client.on_subscribe = on_subscribe
+        mqtt_client.on_unsubscribe = on_unsubscribe
+
+        return mqtt_client
+
+    async def _reconnect_loop(self):
+        logger.debug("Reconnect Daemon starting...")
+        current_attempt = None
+        try:
+            while True:
+                async with self.disconnected_cond:
+                    await self.disconnected_cond.wait_for(
+                        lambda: not self.is_connected() and self._desire_connection
+                    )
+                try:
+                    logger.debug("Reconnect Daemon attempting to reconnect...")
+                    current_attempt = asyncio.create_task(self.connect())
+                    await current_attempt
+                    logger.debug("Reconnect Daemon reconnect attempt succeeded")
+                except exceptions.ConnectionFailedError:
+                    # TODO: determine if exception is retryable
+                    interval = self._reconnect_interval
+                    logger.debug(
+                        "Reconnect Daemon reconnect attempt failed. Trying again in {} seconds".format(
+                            interval
+                        )
+                    )
+                    await asyncio.sleep(interval)
+        except asyncio.CancelledError:
+            logger.debug("Reconnect Daemon was cancelled")
+            current_attempt.cancel()
```

Comment:
`AttributeError: 'NoneType' object has no attribute 'cancel'`

Reply:
Should have seen this one coming tbh

Reply:
Fixed